File: python/loom_engine/ruleset.py
"""loom_engine.ruleset - 5e + PF2e action economy / initiative / conditions.

Byte-parity hand-port of the TypeScript runtime/ruleset.ts. Deterministic +
content-agnostic (condition NAMES are caller-supplied, so no SRD text is
reproduced). Compatible with the D&D 5e SRD (CC-BY-4.0) and the Pathfinder 2e
Remaster ruleset (ORC License) - see ../../NOTICE.md. Pure (no RNG; the d20 is
rolled by the engine PRNG - this does the ruleset-correct ORDERING).
"""

from __future__ import annotations

import functools
from typing import Dict, List, Optional

RULESET_5E = "5e"
RULESET_PF2E = "pf2e"


# ---- action economy ----

def start_turn_budget(ruleset: str) -> dict:
    """5e: 1 action + 1 bonus + 1 reaction. PF2e: 3 actions + 1 reaction."""
    if ruleset == RULESET_PF2E:
        return {"ruleset": RULESET_PF2E, "resources": {"action": 3, "reaction": 1}}
    return {"ruleset": RULESET_5E,
            "resources": {"action": 1, "bonus": 1, "reaction": 1}}


def can_spend(budget: dict, resource: str, n: int = 1) -> bool:
    need = n if isinstance(n, int) and n > 0 else 1
    have = budget.get("resources", {}).get(resource)
    return isinstance(have, int) and have >= need


def spend(budget: dict, resource: str, n: int = 1) -> bool:
    """Spend n (default 1) of a resource. True if spent, False if insufficient."""
    need = n if isinstance(n, int) and n > 0 else 1
    if not can_spend(budget, resource, need):
        return False
    budget["resources"][resource] -= need
    return True


# ---- initiative ordering ----

def _is_pure_numeric(s: str) -> bool:
    """An optional '-' then >=1 ASCII digit."""
    rest = s[1:] if s[:1] == "-" else s
    return len(rest) > 0 and all("0" <= c <= "9" for c in rest)


def _normalize_numeric(s: str):
    neg = s[:1] == "-"
    digits = s[1:] if neg else s
    mag = digits.lstrip("0") or "0"
    return (neg and mag != "0", mag)  # -0 is +0


def _byte_cmp(a: str, b: str) -> int:
    ba = a.encode("utf-8")
    bb = b.encode("utf-8")
    return -1 if ba < bb else (1 if ba > bb else 0)


def compare_ids(a: str, b: str) -> int:
    """Numeric-aware id comparison (Codex P1 / the shadow-wire finding). Numeric
    ids sort by VALUE (2 < 10), strings lexicographically, numbers before strings.
    No int parsing - sign + digit-length + UTF-8 bytes, so ids beyond i64 (uuids,
    huge numbers) are correct. Byte-identical to the TS + Rust comparators."""
    na = _is_pure_numeric(a)
    nb = _is_pure_numeric(b)
    if na and not nb:
        return -1  # numbers before strings
    if not na and nb:
        return 1
    if not na and not nb:
        return _byte_cmp(a, b)
    a_neg, a_mag = _normalize_numeric(a)
    b_neg, b_mag = _normalize_numeric(b)
    if not a_neg and b_neg:
        return 1  # +a > -b
    if a_neg and not b_neg:
        return -1
    if len(a_mag) != len(b_mag):
        mag = -1 if len(a_mag) < len(b_mag) else 1
    else:
        mag = _byte_cmp(a_mag, b_mag)
    by_value = -mag if a_neg else mag  # both negative: larger magnitude is smaller
    if by_value != 0:
        return by_value
    return _byte_cmp(a, b)  # math-equal (e.g. "02" vs "2"): raw bytes, total order


def _initiative_cmp(a: dict, b: dict) -> int:
    for key in ("total", "modifier", "d20"):
        av = int(a.get(key, 0) or 0)
        bv = int(b.get(key, 0) or 0)
        if av != bv:
            return -1 if av > bv else 1  # DESC
    return compare_ids(str(a.get("id", "")), str(b.get("id", "")))


def initiative_order(entries: List[dict]) -> List[dict]:
    """Deterministic order: total DESC, then modifier DESC, then natural d20 DESC,
    then a NUMERIC-AWARE id tiebreak (compare_ids). Correct for BOTH 5e and PF2e
    and for integer ids AND string entity ids. New list; input untouched."""
    return sorted(list(entries), key=functools.cmp_to_key(_initiative_cmp))


# ---- conditions (content-agnostic duration tracker) ----

DURATION_UNTIL_REMOVED = -1


def create_condition_track() -> dict:
    return {"conditions": {}}


def apply_condition(track: dict, condition_id: str, rounds: Optional[int] = None) -> None:
    """Apply / refresh a condition. rounds 0 or None -> 'until removed'."""
    if not condition_id:
        return
    r = int(rounds) if isinstance(rounds, int) else DURATION_UNTIL_REMOVED
    if r == 0:
        r = DURATION_UNTIL_REMOVED
    track["conditions"][condition_id] = r


def remove_condition(track: dict, condition_id: str) -> bool:
    if condition_id in track["conditions"]:
        del track["conditions"][condition_id]
        return True
    return False


def has_condition(track: dict, condition_id: str) -> bool:
    return condition_id in track["conditions"]


def condition_remaining(track: dict, condition_id: str) -> int:
    return track["conditions"].get(condition_id, 0)


def tick_conditions(track: dict) -> List[str]:
    """Tick every FINITE condition down one round; expire (remove) any reaching 0.
    DURATION_UNTIL_REMOVED never ticks. Returns expired ids (insertion order)."""
    conds = track["conditions"]
    expired: List[str] = []
    for cid, rem in list(conds.items()):
        if rem == DURATION_UNTIL_REMOVED:
            continue
        if rem <= 1:
            expired.append(cid)
        else:
            conds[cid] = rem - 1
    for cid in expired:
        del conds[cid]
    # Codex P1: SORTED by UTF-8 bytes so the result matches the Rust core's
    # BTreeMap order - identical across languages, not insertion-dependent.
    expired.sort(key=lambda s: s.encode("utf-8"))
    return expired


def active_conditions(track: dict) -> List[str]:
    """Active condition ids in canonical SORTED order (UTF-8 bytes), matching the
    Rust core's BTreeMap - identical across languages."""
    return sorted(track["conditions"].keys(), key=lambda s: s.encode("utf-8"))


RESOURCE_RULESET = "ruleset"

File: src/runtime/ruleset.ts
// Ruleset Adapters - 5e + PF2e action economy, initiative ordering, conditions.
//
// v2.3.0. Deterministic, content-agnostic primitives for the two rulesets the
// engine targets. PURE: no RNG (the d20 itself is rolled by the engine's seeded
// PRNG; this module does the ruleset-correct ORDERING), no wall-clock. It covers
// the three mechanics that differ by system: the per-turn action budget, the
// initiative tiebreak, and a condition-duration tracker.
//
// Compatible with the D&D 5e SRD (CC-BY-4.0) and the Pathfinder Second Edition
// Remaster ruleset (ORC License) - see NOTICE.md. Not affiliated with or
// endorsed by Wizards of the Coast or Paizo. Condition NAMES are supplied by the
// caller (the tracker is content-agnostic), so no SRD text is reproduced here.
//
// Code style: var-only, no arrow functions.

export type RulesetId = '5e' | 'pf2e';

// ---- Action economy -------------------------------------------------------

export interface TurnBudget {
  ruleset: RulesetId;
  resources: { [resource: string]: number };
}

// The per-turn starting budget. 5e: 1 action + 1 bonus + 1 reaction. PF2e:
// 3 actions + 1 reaction. (Movement is tracked by the positioning layer and the
// reaction by the reaction-economy module across turns; this is the ACTIVE
// in-turn budget.)
export function startTurnBudget(ruleset: RulesetId): TurnBudget {
  if (ruleset === 'pf2e') {
    return { ruleset: 'pf2e', resources: { action: 3, reaction: 1 } };
  }
  return { ruleset: '5e', resources: { action: 1, bonus: 1, reaction: 1 } };
}

// True iff at least `n` (default 1) of `resource` remains.
export function canSpend(budget: TurnBudget, resource: string, n?: number): boolean {
  var need = typeof n === 'number' && n > 0 ? n : 1;
  var have = budget.resources[resource];
  return typeof have === 'number' && have >= need;
}

// Spend `n` (default 1) of a resource. Returns true if spent, false if
// insufficient (no change on failure).
export function spend(budget: TurnBudget, resource: string, n?: number): boolean {
  var need = typeof n === 'number' && n > 0 ? n : 1;
  if (!canSpend(budget, resource, need)) return false;
  budget.resources[resource] = (budget.resources[resource] as number) - need;
  return true;
}

// ---- Initiative ordering --------------------------------------------------

export interface InitiativeEntry {
  id: string;
  total: number;       // d20 + modifier
  modifier?: number;   // DEX mod (5e) / initiative mod (PF2e)
  d20?: number;        // the natural roll, for tiebreaks
}

// Deterministic initiative order: total DESC, then modifier DESC, then natural
// d20 DESC, then id ASC (stable). One tiebreak for BOTH 5e and PF2e (both break
// initiative ties by the modifier). Returns a NEW sorted array; input untouched.
export function initiativeOrder(entries: ReadonlyArray<InitiativeEntry>): InitiativeEntry[] {
  var copy = entries.slice();
  copy.sort(function (a, b) {
    if (b.total !== a.total) return b.total - a.total;
    var am = a.modifier || 0;
    var bm = b.modifier || 0;
    if (bm !== am) return bm - am;
    var ad = a.d20 || 0;
    var bd = b.d20 || 0;
    if (bd !== ad) return bd - ad;
    return a.id < b.id ? -1 : (a.id > b.id ? 1 : 0);
  });
  return copy;
}

// ---- Conditions (content-agnostic duration tracker) -----------------------

// A condition's remaining duration: a positive round count, or this sentinel
// for "until removed" (never ticks down). Covers both 5e "for N rounds" and
// PF2e value-N conditions as well as open-ended ones.
export const DURATION_UNTIL_REMOVED = -1;

export interface ConditionTrack {
  conditions: Map<string, number>;
}

export function createConditionTrack(): ConditionTrack {
  return { conditions: new Map() };
}

// Apply (or refresh) a condition by id. A `rounds` of 0 or omitted is treated as
// "until removed"; a positive value is a finite duration.
export function applyCondition(track: ConditionTrack, conditionId: string, rounds?: number): void {
  if (!conditionId) return;
  var r = typeof rounds === 'number' ? Math.floor(rounds) : DURATION_UNTIL_REMOVED;
  if (r === 0) r = DURATION_UNTIL_REMOVED;
  track.conditions.set(conditionId, r);
}

export function removeCondition(track: ConditionTrack, conditionId: string): boolean {
  return track.conditions.delete(conditionId);
}

export function hasCondition(track: ConditionTrack, conditionId: string): boolean {
  return track.conditions.has(conditionId);
}

// Remaining rounds for a condition, DURATION_UNTIL_REMOVED if open-ended, or 0
// if absent.
export function conditionRemaining(track: ConditionTrack, conditionId: string): number {
  var v = track.conditions.get(conditionId);
  return v === undefined ? 0 : v;
}

// Tick every FINITE condition down one round; expire (remove) any that reach 0.
// The DURATION_UNTIL_REMOVED sentinel never ticks. Returns the expired ids in
// insertion order (deterministic).
export function tickConditions(track: ConditionTrack): string[] {
  var expired: string[] = [];
  for (var entry of track.conditions) {
    var id = entry[0];
    var rem = entry[1];
    if (rem === DURATION_UNTIL_REMOVED) continue;
    if (rem <= 1) expired.push(id);
    else track.conditions.set(id, rem - 1);
  }
  for (var e of expired) track.conditions.delete(e);
  return expired;
}

// Active condition ids in insertion order (deterministic).
export function activeConditions(track: ConditionTrack): string[] {
  var out: string[] = [];
  for (var entry of track.conditions) out.push(entry[0]);
  return out;
}

export const RESOURCE_RULESET = 'ruleset';
